File: src/main/java/ua/softgroup/matrix/server/service/impl/UserServiceImpl.java
package ua.softgroup.matrix.server.service.impl;

import org.mindrot.jbcrypt.BCrypt;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import retrofit2.Response;
import ua.softgroup.matrix.server.desktop.model.datamodels.AuthModel;
import ua.softgroup.matrix.server.persistent.entity.User;
import ua.softgroup.matrix.server.persistent.repository.UserRepository;
import ua.softgroup.matrix.server.service.UserService;
import ua.softgroup.matrix.server.supervisor.consumer.endpoint.SupervisorEndpoint;
import ua.softgroup.matrix.server.supervisor.consumer.json.LoginJson;
import ua.softgroup.matrix.server.supervisor.consumer.json.UserJson;

import javax.validation.Validator;
import java.io.IOException;
import java.util.Optional;

@Service
public class UserServiceImpl extends AbstractEntityTransactionalService<User> implements UserService {

    private static final Logger LOG = LoggerFactory.getLogger(UserServiceImpl.class);

    private final Validator validator;
    private final SupervisorEndpoint supervisorEndpoint;

    @Autowired
    public UserServiceImpl(UserRepository repository, Validator validator, SupervisorEndpoint supervisorEndpoint) {
        super(repository);
        this.validator = validator;
        this.supervisorEndpoint = supervisorEndpoint;
    }

    @Override
    public String authenticate(AuthModel authModel) {
//        if (!validator.validate(up).isEmpty()) return Constants.INVALID_USERNAME.toString();
        try {
            return tryToAuthenticate(authModel.getUsername(), authModel.getPassword());
        } catch (IOException e) {
            LOG.warn("Authentication error: {}", e);
            return authenticateFromDb(authModel.getUsername(), authModel.getPassword());
        }
    }

    private String authenticateFromDb(String username, String password) {
        User user = getRepository().findByUsername(username);
        if (user != null && BCrypt.checkpw(password, user.getPassword())) {
            LOG.info("Offline authentication. Given token {}", user.getTrackerToken());
            return user.getTrackerToken();
        }
        return null;
    }

    private String tryToAuthenticate(String username, String password) throws IOException {
        LoginJson loginJson = executeLoginQuery(username, password).body();
        if (loginJson.getSuccess()) {
            saveUser(loginJson.getUser(), password);
            String token = loginJson.getTrackerToken();
            LOG.info("User authenticated successfully: {}", token);
            return token;
        }
        LOG.info("Authentication failed: {}", loginJson.getMessage());
        return null;
    }

    private Response<LoginJson> executeLoginQuery(String username, String password) throws IOException {
        Response<LoginJson> response = supervisorEndpoint
                .login(username, password)
                .execute();
        if (!response.isSuccessful()) {
            throw new IOException("Oops... Something goes wrong. " + response.errorBody().string());
        }
        return response;
    }

    private void saveUser(UserJson userJson, String password) {
        LOG.debug("UserJson {}", userJson);
        User user = Optional.ofNullable(getRepository().findOne(userJson.getId())).orElse(new User());
        LOG.debug("UserEntity {}", user);
        user.setId(userJson.getId());
        user.setPassword(BCrypt.hashpw(password, BCrypt.gensalt()));
        user.setUsername(userJson.getUsername());
        user.setTrackerToken(userJson.getTrackerToken());
        user.setExternalHourlyRate(userJson.getProfile().getExternalHourlyRate());
        user.setExternalHourlyRateCurrencyId(userJson.getProfile().getExternalHourlyRateCurrencyId());
        user.setInternalHourlyRate(userJson.getProfile().getInternalHourlyRate());
        user.setInternalHourlyRateCurrencyId(userJson.getProfile().getInternalHourlyRateCurrencyId());

        getRepository().save(user);
    }

    @Override
    public Optional<User> getByUsername(String username) {
        return Optional.ofNullable(getRepository().findByUsername(username));
    }

    @Override
    public Optional<User> getByTrackerToken(String token) {
        return Optional.ofNullable(getRepository().findByTrackerToken(token));
    }

    @Override
    protected UserRepository getRepository() {
        return (UserRepository) repository;
    }

}

File: src/main/scala/ua/softgroup/matrix/server/service/impl/UserServiceImpl.scala
package ua.softgroup.matrix.server.service.impl

import java.io.IOException
import java.util.Optional

import org.mindrot.jbcrypt.BCrypt
import org.slf4j.LoggerFactory
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Service
import retrofit2.Response
import ua.softgroup.matrix.api.model.datamodels.AuthModel
import ua.softgroup.matrix.server.persistent.entity.User
import ua.softgroup.matrix.server.persistent.repository.UserRepository
import ua.softgroup.matrix.server.service.UserService
import ua.softgroup.matrix.server.supervisor.consumer.endpoint.SupervisorEndpoint
import ua.softgroup.matrix.server.supervisor.consumer.json.{LoginJson, UserJson}

import scala.util.{Failure, Success, Try}

/**
  * @author Oleksandr Tyshkovets <sg.olexander@gmail.com>
  */
@Service
class UserServiceImpl @Autowired() (repository: UserRepository,
                                    supervisorEndpoint: SupervisorEndpoint) extends UserService {

  private val logger = LoggerFactory.getLogger(this.getClass)

  override def authenticate(authModel: AuthModel): Option[String] = {
    val username = authModel.getUsername
    val password = authModel.getPassword

    Try(authenticate(username, password)) match {
      case Success(optionToken) => optionToken
      case Failure(f) =>
        logger.error("Authentication error", f.getCause)
        authenticateFromDb(username, password)
    }
  }

  private def authenticateFromDb(username: String, password: String): Option[String] = {
    val user = repository.findByUsername(username)
    if (user != null && BCrypt.checkpw(password, user.getPassword)) {
      logger.info(s"Offline authentication. User '$username' authenticated successfully with token '${user.getTrackerToken}'")
      Option(user.getTrackerToken)
    }
    logger.info(s"Offline authentication. User '$username' not found or passwords aren't equals")
    Option.empty
  }

  private def authenticate(username: String, password: String): Option[String] = {
    val loginJson = executeLoginQuery(username, password).body
    if (loginJson.success) {
      saveUser(loginJson.user, password)
      val token = loginJson.trackerToken
      logger.info(s"User '$username' authenticated successfully with token '$token'")
      return Option(token)
    }
    logger.info(s"Authentication failed: ${loginJson.message}")
    Option.empty
  }

  private def executeLoginQuery(username: String, password: String): Response[LoginJson] = {
    val response = supervisorEndpoint
      .login(username, password)
      .execute
    if (!response.isSuccessful) {
      throw new IOException(s"Oops... Something goes wrong. ${response.errorBody.string}")
    }
    response
  }

  private def saveUser(userJson: UserJson, password: String) = {
    logger.debug(s"UserJson $userJson")
    val user = Optional.ofNullable(repository.findOne(userJson.id)).orElse(new User)
    logger.debug(s"UserEntity $user")
    user.setId(userJson.id)
    user.setPassword(BCrypt.hashpw(password, BCrypt.gensalt))
    user.setUsername(userJson.username)
    user.setTrackerToken(userJson.trackerToken)
    user.setExternalHourlyRate(userJson.profile.externalHourlyRate)
    user.setExternalHourlyRateCurrencyId(userJson.profile.externalHourlyRateCurrencyId)
    user.setInternalHourlyRate(userJson.profile.internalHourlyRate)
    user.setInternalHourlyRateCurrencyId(userJson.profile.internalHourlyRateCurrencyId)

    repository.save(user)
  }

  override def getByUsername(username: String): Option[User] = Option(repository.findByUsername(username))

  override def getByTrackerToken(token: String): Optional[User] = Optional.ofNullable(repository.findByTrackerToken(token))

  override def getById(id: Long): Optional[User] = Optional.ofNullable(repository.findOne(id))

  override def save(entity: User): User = repository.save(entity)

  override def isExist(id: Long): Boolean = repository.exists(id)

}